### ISSUE TITLE
RFC: Inconsistent outbound passthrough behavior if source or dest workloads are not found in the workload API

### ISSUE THREAD

**Issue by bleggett** [istio/ztunnel#369]
We've found a bug where a pod that makes outbound calls the k8s API as part of its readiness checks gets stuck due to ztunnel dropping the connection, leading to the pod never entering readiness.

This is exposed by the fact that ztunnel drops the connection if it cannot find the _source_ workload in the xDS API - and indeed the source workload is not present in the xDS API from ztunnel's perspective.

That in and of itself may be a separate bug with how and when pods are added to the workload registry that isnt' hit in the sidecar model (because I don't think sidecars drop passthru traffic if they can't find their _own_ registration, that seems pointless), but raises another question:

In https://github.com/istio/ztunnel/blob/master/src/proxy/outbound.rs#L303

- If the `source` is not found in the workload API, we drop the connection.
- If the `dest` is not found in the workload API, we give up and  pass the connection thru anyway.

* Why are we stricter with `source` than `dest`? 
* Should we pass thru if we cannot find `source` just like we do `dest`? This would functionally mimic the existing sidecar model, but I actually think there are benefits to being strict on  this point for both in `ztunnel`, as it helps expose consistency issues.
* Should ztunnel's strictness here be maybe influenced by e.g. `outboundTrafficPolicy.mode = REGISTRY_ONLY`?

My inclination is that ultimately ztunnel should probably respect `outboundTrafficPolicy` here.

There is also an argument to be made that for now we should passthru outbound traffic if we cannot find  the `source`, just as we do for `dest`, as this mirrors current behavior.

**Comment by howardjohn** [comment]
Can you clarify why source is missing? If its because we don't send the Workload until its ready that is a bug - we should send those with some signal they are not healthy. This is important for outbound as well - you should be able to reach a not ready pod if you specifically send to its IP, it just shouldn't be part of a VIP

**Comment by bleggett** [comment]
@howardjohn Yeah that's what I'm doing - I think that is the bug, and we simply haven't hit it because this kind of order of operations doesn't happen in sidecar mode.

Unless I encounter a good reason why the pod _can't_ be added to the workload API in an unready state, my pref would be to do that and keep the strict `source` check here.

**Comment by howardjohn** [comment]
We should probably make it an enum as well as k8s has some special behavior around TERMINATING etc that we will want to handle at some point

**Comment by bleggett** [comment]
The workload xDS registration bug should be resolved by https://github.com/istio/istio/pull/43085 - but leaving this issue open for tracking as I believe the question of how strict ztunnel should be for `source` or `dest` and whether that is configurable or potentially affected by `outboundTrafficPolicy` is still TBD.

**Comment by howardjohn** [comment]
So dest unknown by default definitely must be allowed by default, or all external (google.com) traffic is broken. If we tie it to outboundTrafficPolicy, everything would still be broken, since we cannot even match a SE for "google.com" at L4 (unless they define a VIP, which is unlikely for most sites). This means that 100% of the mesh would need an egress waypoint, which is counter to our goals. In general, even for sidecars, I think REGISTRY_ONLY is a poor policy, one users often confuse as a security policy ("egress firewall") when it absolutely is NOT an egress firewall. I have advocated for completely removing it in the past. So I would prefer to not implement it in ztunnel for those reasons.

For source unknown - we _could_ allow it if the upstream is not HBONE. If it is HBONE then we cannot though, as we don't know which certificate to attach. I think having a pod legitimately send outbound traffic and have source unknown represents a bug though, so it seems safer to fail closed in the event of a bug. There was an issue about blocking pod startup until the ztunnel has the source info which would make this not possible unless something went wrong; without that we just have a small race condition.

**Comment by hzxuzhonghu** [comment]
How about the case that istio controlplane has not synced the pod before the pod sends out a request? Will this kind of request being intercepted to ztunnel?

**Comment by howardjohn** [comment]
https://github.com/istio/istio/issues/42850 is about syncing

**Comment by bleggett** [comment]
> So dest unknown by default definitely must be allowed by default, or all external (google.com) traffic is broken. If we tie it to outboundTrafficPolicy, everything would still be broken, since we cannot even match a SE for "google.com" at L4 (unless they define a VIP, which is unlikely for most sites). This means that 100% of the mesh would need an egress waypoint, which is counter to our goals. In general, even for sidecars, I think REGISTRY_ONLY is a poor policy, one users often confuse as a security policy ("egress firewall") when it absolutely is NOT an egress firewall. I have advocated for completely removing it in the past. So I would prefer to not implement it in ztunnel for those reasons.

Inclined to agree. `REGISTRY_ONLY` is IMO largely redundant with/sort of pointless in the context of e.g. STRICT mode or similar anyway, if you actually care about restricting traffic to be exclusively within the mesh.

> I think having a pod legitimately send outbound traffic and have source unknown represents a bug though, so it seems safer to fail closed in the event of a bug. There was an issue about blocking pod startup until the ztunnel has the source info which would make this not possible unless something went wrong; without that we just have a small race condition.

Yeah at this point I'm comfortable saying that source unknown is a bug - if the pod has an IP, it should be in the source registry and ztunnel should see it. If it does not yet, ztunnel shouldn't see it, and we shouldn't have a problem. In general the possibility space for startup+network availability bugs in ambient is IMO much smaller anyway due to the lack of sidecars.

**Comment by bleggett** [comment]
> How about the case that istio controlplane has not synced the pod before the pod sends out a request? Will this kind of request being intercepted to ztunnel?

ztunnel will drop this, correct. If the pod sends out a single request as part of it's liveness/readiness checks and dies, and does not implement retries, this could still be a problem.

This would be both uncommmon and rather difficult to fix on our side, and I think the recommendation would be to add retries to whatever external calls your readiness/liveness checks are making, as a matter of basic hygiene.

**Comment by bleggett** [comment]
Aight closing this as I think there's consensus that current behavior is reasonable for reasons laid out here, and ATM there is no anticipated need to change it. If later we turn out to be wrong, we can reopen.
